### PR TITLE
Autocomplete for mentions

### PR DESCRIPTION
**Pull request by nickvergessen**
- [ ] ~~Handle guests~~ => Does not work, because they have no unique user id
- [x] Add UI
- [x] Document endpoint
- [x] Sort results (by last chat message or alphabetic? @jancborchardt )
- [x] Make ISearchResult injectable in the server => https://github.com/nextcloud/server/pull/9310

> bildschirmfoto von 2018-04-26 13-02-33

### CODE REVIEW

**Comment by jancborchardt** [comment]
Sorting should be by best match rather than alphabetic. Last chat message is already a good metric for "best match". :)

**Comment by nickvergessen** [comment]
@danxuliu Can you help copying the UI/UX changes from https://github.com/nextcloud/server/pull/6982/files

File: js/views/chatview.js [hunk 1]
```diff
@@ -98,6 +98,79 @@
...
+			}
+			if(!_.isUndefined(this._autoCompleteRequestTimer)) {
+				clearTimeout(this._autoCompleteRequestTimer);
+			}
+			this._autoCompleteRequestTimer = _.delay(function() {
+				if(!_.isUndefined(this._autoCompleteRequestCall)) {
+					this._autoCompleteRequestCall.abort();
+				}
+				this._autoCompleteRequestCall = $.ajax({
+					url: OC.linkToOCS('apps/spreed/api/v1/chat', 2) + 'dshwd9uv' /* FIXME this.token */ + '/autocomplete',
```
Comment (nickvergessen) [inline js/views/chatview.js]: :see_no_evil: we don't seem to have access to the room info in this view?

**Comment by danxuliu** [comment]
@nickvergessen For consistency with the Comments app, [the author should not be suggested in the autocompletion](https://github.com/nextcloud/server/pull/7914/commits/6fdd6861118f55a9e740552b82382f55e9415eea). Also it seems that matching by display name is case sensitive; it should be case insensitive.

**Comment by danxuliu** [comment]
@nickvergessen Unless I have missed something the UI should be ready.

**Comment by danxuliu** [comment]
I have rebased on _master_ and amended the CSS file to solve the "conflict" with #814 (it was not really a conflict, but the CSS rules were not properly sorted due to the automatic merging).

**Pull request merged**